Update WORKFLOW.md to be more descriptive (#785)
hoprnet/hoprnet: WORKFLOW.md

# Development workflow

Principles:
## Introduction

- Rely on automation as much as possible.
- Automated tests should prevent bad deployments. Continuous integration,
  continuous deployment.
The HOPR Association team members have agreed on the following development workflow to streamline the process we use to implement the HOPR protocol.

## Goal Workflow
## Principles

- A team-member, or an external contributor writes code in a feature branch, or
  in their own fork.
- **Automation-first**: Rely on automation as much as possible.
- **Tests & CI**: Tests should prevent bad deployments and API regressions. Continuous integration ensures all our code is tested before being merged into our baseline.
- **Releases & CD**: Every week, we do a code-freeze in our codebase by branching out a specific release, which we then deploy on every change.

- The PR should track master
## Rules

> - They create a draft pull-request while they work on the code so we can stay

Reviewer: Let's leave draft PR's in.

Author: I mentioned draft below by “Daily Development”, just moved it to a new place.

> - All PR‘s have to be approved by two team members.

Reviewer: 1?

Author: Poor phrasing, changing it to `All PR‘s have to be approved by an team member different from the one who created the PR.`

> - `master` tracks `release/**`, but releases do not track `master`. i.e. we always merge `release/**` back to `master`, but we do not merge `master` back to a frozen `release`.

Reviewer: Master doesn't track anything.

Releases _can_ be merged back to master, but not always necessary.

Author: Gotcha, replaced the last one with `Releases can be merged back to master, but not always necessary.`

> 1. A team-member or an external contributor writes code in a **feature branch**, or in their own fork. The feature branch should track an **issue**, and describe its role (e.g. “fixes 341")

Reviewer: ```suggestion
1. A team-member or an external contributor writes code in a **feature branch**, or in their own fork. The feature branch describe its goal, and any ticket it is related to. (e.g. “fixes 341")
```

> 2. Immediately after the creation of the branch, create a draft pull-request to stay up to date on progress. That way, every push on will build the project and runs the tests.

Reviewer: Doesn't need to be immediately, but should be done early.

3. When the code is ready, mark the pull-request as ready to review, and request a code-review from a maintainer. In the case of team members, the approval of an additional team member is required. Externals require **two** maintainers to approve the change.

4. After a different team member reviews the code and indicates approval, the PR can be merged. If the history is messy, the PR can be squashed, otherwise, it is merged.

### Release Cycle

1. On Thursdays at 12 pm CEST, the PM Lead of the week will code-freeze `master` by creating a `release/**` branch tracking `master`. Release specific changes will be done in this branch to trigger this particular release then.

> 2. **(TEMPORAL, see #709)** After the `release/**` action has completed, we then crease a `cd/**` branch tracking `release/**`.

Reviewer: Let's flag as TEMPORARY, not TEMPORAL.

3. The instructions with the release, how to test and what bootstrap server to use, are then shared in our social media channels.

## Actions

We made active use of actions to automate tasks trivial to our workflow.

> - **Patch Version Bump**. When a PR to `master` is merged, an action bumps the package.json by a `patch` and commits that change to `master`.

Reviewer: No - it bumps the prerelease tag.

Author: Thanks, good call

> - **Tag Release**. When a PR to `master` is merged, a tag is pushed specifying that feature on that version.

Reviewer: No, does not happen.

Author: @peterbraden how are we pushing all these tags then?

Reviewer: I'm wrong actually.

> - **Release Version Bump**. When a `release/**` branch is pushed with a new commit, an action bumps the package.json by a `minor` and commits that change to the `release` branch. Further pushes on `release` branches would then bump the version by a `patch`.

Reviewer: On the first release build, it bumps minor (clears the prerelease tag).

Subsequent commits bump the patch.

Author: Thanks, rephrased it.

## Branches

- `master`: In our case, `master` is a **prerelease** branch - tests _must_ pass, and code _should_ be stable, but its _acceptable_ to have issues.

- `release/**`: Every week on Thursdays at 12 pm CEST, we cut a `release/**` branch, using an internal name to identify the release.

> - `hotfix/**/**`: Hotfixes to be PR's to release branches. The release branch should then be pulled back into master.

Reviewer: we don't need a hotfix branch format - hotfixes are just branches, and PR to release.

Author: Fine, let's tho try and bring some standard as we are discussing in #656

> - `cd/**` **TEMPORAL, see (#709)**: After a successful release, we then branch out a `cd` branch which will deploy the artefacts based on the release builds from our CI.

Reviewer: s/TEMPORAL/TEMPORARY/
